src/cpp/server/server_context.cc

```diff
@@ -174,33 +174,40 @@ void ServerContextBase::CompletionOp::FillOps(internal::Call* call) {
 }
 
 bool ServerContextBase::CompletionOp::FinalizeResult(void** tag, bool* status) {
-  // Decide whether to call the cancel callback within the lock
-  bool call_cancel;
+  // Decide whether to do the unref or call the cancel callback within the lock
+  bool do_unref = false;
+  bool has_tag = false;
+  bool call_cancel = false;
 
   {
     grpc_core::MutexLock lock(&mu_);
     if (done_intercepting_) {
       // We are done intercepting.
-      bool has_tag = has_tag_;
+      has_tag = has_tag_;
       if (has_tag) {
         *tag = tag_;
       }
-      Unref();
-      return has_tag;
-    }
-    finalized_ = true;
+      // Release the lock before unreffing as Unref may delete this object
+      do_unref = true;
+    } else {
+      finalized_ = true;
+
+      // If for some reason the incoming status is false, mark that as a
+      // cancellation.
+      // TODO(vjpai): does this ever happen?
+      if (!*status) {
+        cancelled_ = 1;
+      }
 
-    // If for some reason the incoming status is false, mark that as a
-    // cancellation.
-    // TODO(vjpai): does this ever happen?
-    if (!*status) {
-      cancelled_ = 1;
+      call_cancel = (cancelled_ != 0);
+      // Release the lock since we may call a callback and interceptors.
     }
-
-    call_cancel = (cancelled_ != 0);
-    // Release the lock since we may call a callback and interceptors.
   }
 
+  if (do_unref) {
+    Unref();
```

Comment:
nit: Internally, we usually add a comment here to avoid similar issues in the future:

```
// `Unref()` can delete `this`. So, do not access `this` afterwards.
```

Reply:
Thanks, I've gone ahead and done this. I guess internally Unref would also be under LOCKS_EXCLUDED, correct? Maybe we can do that in this codebase soon. @veblush - are absl lock annotations available to us?